### ISSUE TITLE
Low regression loss during Training

### ISSUE THREAD

**Issue by awilliamson** [fizyr/keras-retinanet#184]
When training, either coco, or pascal, the regression loss always seems to initialise as 0.00X or thereabouts. This seems awfully small, and given that the overall loss function is just the sum of both loss scores, it would seems that the regressor would not sufficiently learn the task.

I'm not quite sure if this is expected behaviour, or whether it's a sign of something deeper. If anybody has an insight into this, that would be great. I'm simply concerned how a loss function can be so low ( indicating it is solved almost ) when it hasn't even trained for 1 epoch of even a few samples. Especially given the amount bboxes are off in final visualisations.

**Comment by hgaiser** [comment]
Did you update recently? I changed the division for the regression loss not so long ago.

**Comment by awilliamson** [comment]
I do remember you talking about the L1 loss change! The original retinanet paper seems to defer to standard two-stage detector mechanisms where smooth l1 loss is used for bbox regressors. According to the Faster RCNN paper the bbox regression uses smooth loss where regression loss is only activated for positive anchors. ( paper in front of me ).
So following from our Slack discussion, it would appear this is the correct form of the loss :)

Indeed, under new L1 Loss we have whopping >4 regression loss initially!

Interestingly what is the ratio of negative anchors to positive? I recall a ratio of 1:3 respectively being defined in some papers; not sure if that holds for RetinaNet from their paper however.

**Comment by hgaiser** [comment]
This issue can be considered resolved then?

**Comment by awilliamson** [comment]
Not quite sure to be honest. Empirical results from yourself, and my testing shows division by just positive anchors to be correct. Intuitively this makes sense. If only positive bbox counts to the Loss, then it should only be normalised by those elements included in the summation, right?

I can't understand why the Faster R-CNN paper states it as normalisation by all anchors, when only positive counts towards it. How their code even works with this. And more importantly why we discovered loss of 0.004 vs 4.0. The only thing I can think of is the value of lambda, to weight the bbox regression loss in the overall loss metric. This might explain different empirical results.

I do think more discussion and discovery is required around this dichotomy to see which side of the fence we should land on to be correct.

**Comment by hgaiser** [comment]
I agree, but I'm not sure if it's worthwhile to have this issue open, as there isn't a direct action we can do to resolve this (is there even something to resolve?). I'm closing this for now, but feel free to reopen it if there is something actionable to do.
